resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KubernetesFeatureConfigStep.scala

```diff
@@ -18,13 +18,17 @@ package org.apache.spark.deploy.k8s.features
 
 import io.fabric8.kubernetes.api.model.HasMetadata
 
+import org.apache.spark.annotation.DeveloperApi
 import org.apache.spark.deploy.k8s.SparkPod
 
 /**
+ * :: DeveloperApi ::
+ *
  * A collection of functions that together represent a "feature" in pods that are launched for
  * Spark drivers and executors.
  */
-private[spark] trait KubernetesFeatureConfigStep {
+@DeveloperApi
+trait KubernetesFeatureConfigStep {
```

Comment:
FYI, we should [avoid using trait for Java APIs](https://github.com/databricks/scala-style-guide#java-traits). I would either mark it with `@Unstable` or make it abstract class.

Reply:
Also we shouldn't rely on Scala collections for Java users, for example, at `getAdditionalPodSystemProperties` and `getAdditionalKubernetesResources`.

There was a bit of discussions (with @tgravescs) about retuning a Java map vs having another method to return Java (e.g. `getAdditionalPodSystemPropertiesJava`) . Both ways are already being used in the current codebase, and we should better have one way to do it. Which way do you prefer?

Reply:
If we are making this a pluggable api it would be better to make this a java interface like the ExecutorPlugin, DrivePlugin, ResourceDiscoveryPlugin and LocalDiskShuffleDataIO.

Reply:
So given it's a developer API I think it might be OK for it to be a Scala trait.

Reply:
Even for a developer API, we should make it Scala/Java friendly. That's why I suggested to make it `@Unstable` (and make it better next time), or properly consider Java interoperability. Also, using traits sometimes bring some surprises when it comes to binary compatibility.

Reply:
Unstable sounds good @HyukjinKwon :)

Reply:
that sounds fine, note by comment below about SparkPod still being private[spark] though, that is used by this api so should also be opened up.